pkg/reconciler/route/reconcile_resources.go

```diff
@@ -173,30 +215,86 @@ func (c *Reconciler) reconcilePlaceholderServices(ctx context.Context, route *v1
 	return services, nil
 }
 
-func (c *Reconciler) updatePlaceholderServices(ctx context.Context, route *v1.Route, services []*corev1.Service, ingress *netv1alpha1.Ingress) error {
+func (c *Reconciler) updatePlaceholderServices(ctx context.Context, route *v1.Route, pairs []resources.ServicePair, ingress *netv1alpha1.Ingress) error {
 	logger := logging.FromContext(ctx)
 	ns := route.Namespace
 
 	eg, egCtx := errgroup.WithContext(ctx)
-	for _, service := range services {
-		service := service
+	for _, pair := range pairs {
+		from := pair
 		eg.Go(func() error {
-			desiredService, err := resources.MakeK8sService(egCtx, route, service.Name, ingress, resources.IsClusterLocalService(service), service.Spec.ClusterIP)
+			to, err := resources.MakeK8sService(egCtx, route, from.Tag, ingress, resources.IsClusterLocalService(from.Service))
 			if err != nil {
 				// Loadbalancer not ready, no need to update.
 				logger.Warnw("Failed to update k8s service", zap.Error(err))
 				return nil
 			}
 
-			// Make sure that the service has the proper specification.
-			if !equality.Semantic.DeepEqual(service.Spec, desiredService.Spec) {
-				// Don't modify the informers copy.
-				existing := service.DeepCopy()
-				existing.Spec = desiredService.Spec
-				if _, err := c.kubeclient.CoreV1().Services(ns).Update(ctx, existing, metav1.UpdateOptions{}); err != nil {
+			canUpdate := false
+
+			if from.Spec.Type != to.Spec.Type {
+				switch from.Spec.Type {
+				// Transitions from ExternalName to any type should work
+				case corev1.ServiceTypeExternalName:
+					canUpdate = true
+				default:
+					// Transitions from ClusterIP to ExternalName Fail
+					// See: https://github.com/kubernetes/kubernetes/issues/104329
+				}
+			} else if from.Spec.Type == corev1.ServiceTypeClusterIP {
+				if from.Spec.ClusterIP != to.Spec.ClusterIP {
+					if from.Spec.ClusterIP != corev1.ClusterIPNone && to.Spec.ClusterIP == "" {
+						// Copy over the cluster IP
+						to.Spec.ClusterIP = from.Spec.ClusterIP
+						canUpdate = true
+					}
```

Comment:
the implied else case of this inner if will delete and re-create the service, is that what we want? (The default branch of the switch above explains needing to delete/create when going from ClusterIP to ExternalName, but it's not clear why or if we need a delete/create in this case?)

Reply:
clusterIP is immutable so `None` to `` (blank) requires a recreate unfortunately.

Reply:
I'll clean up `if` and add a comment